### PR TITLE
Attachment uploads

### PR DESCRIPTION
**Pull request by timoballard**
[Link to ZenHub issue.](https://app.zenhub.com/workspaces/doj-crt-intake-scrum-board-5d03bf56c1c8a35d482eca0f/issues/18f/crt-portal/41)

## What does this change?
Adds support for attaching files to existing reports. Files are stored in a private S3 bucket. Support for [downloading](https://app.zenhub.com/workspaces/doj-crt-intake-scrum-board-5d03bf56c1c8a35d482eca0f/issues/18f/crt-portal/884) and [removing](https://app.zenhub.com/workspaces/doj-crt-intake-scrum-board-5d03bf56c1c8a35d482eca0f/issues/18f/crt-portal/885) files will be added separately. 

## Screenshots (for front-end PR):
image
image
image
image
image

### Author

+ [x] If this is a story, run locally and check to make sure all Acceptance Criteria are met. (If any criteria are unclear, ask about them.).
+ [x] Check for, document, and establish a testing plan for any behavior that may vary across environments or is otherwise difficult to test.
+ [x] Check for [accessibility](/docs/a11y_plan.md).
+ [x] [Tests pass](https://github.com/USDOJ/crt-portal/#tests).

### Reviewer

+ [ ] If this is a story, run locally and check to make sure all Acceptance Criteria are met. (If any criteria are unclear, ask about them.).
+ [ ] Check for any behavior that may vary across environments or is difficult to test, and ensure that it is well-understood, documented, and that there is a testing plan in place.
+ [ ] Re-check for [accessibility](/docs/a11y_plan.md).
+ [ ] [Tests pass](https://github.com/USDOJ/crt-portal/#tests).

## Notes for reviewer:

See [PR instructions doc](https://github.com/usdoj/crt-portal/blob/master/docs/pull_requests.md) for full pull request review instructions.

### CODE REVIEW

File: crt_portal/cts_forms/migrations/0100_create_reportattachments_20210209_1609.py [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+        migrations.AlterField(
+            model_name='report',
+            name='last_incident_year',
+            field=models.PositiveIntegerField(blank=True, null=True, verbose_name=django.core.validators.MaxValueValidator(2021, message='Date can not be in the future.')),
+        ),
+        migrations.AlterField(
+            model_name='report',
+            name='primary_complaint',
+            field=models.CharField(choices=[('voting', 'Voting rights or ability to vote affected'), ('workplace', 'Workplace discrimination or other employment-related problem'), ('housing', 'Housing discrimination or harassment'), ('education', 'Discrimination at a school, educational program or service, or related to receiving education'), ('police', 'Mistreated by police, correctional staff, or inmates'), ('commercial_or_public', 'Discriminated against in a commercial location or public place'), ('something_else', 'Something else happened')], default='', max_length=100),
+        ),
```
Comment (timoballard) [inline crt_portal/cts_forms/migrations/0100_create_reportattachments_20210209_1609.py:46]: These are unrelated to this card but showed up when running `makemigrations`

Comment (Jkrzy) [reply crt_portal/cts_forms/migrations/0100_create_reportattachments_20210209_1609.py:46]: Thanks for catching these

**Comment by Jkrzy** [comment]
Working through this, looking great! A bit of additional docs on creating the required buckets for local testing would be helpful.

File: docker-compose.yml [hunk 2]
```diff
@@ -5,11 +5,17 @@ services:
     image: postgres:10.1-alpine
     volumes:
       - postgres_data:/var/lib/postgresql/data/
+    ports:
+      - 5432:5432
```
Comment (Jkrzy) [inline docker-compose.yml:9]: Tagging as a potential remnant from development, doesn't look related to other changes.

File: crt_portal/cts_forms/validators.py [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+    'INFECTED': [406],
+    'ERROR': [400, 412, 500, 501],
+}
+
+
+def _scan_file(file):
+    return requests.post(settings.AV_SCAN_URL, files={'file': file}, data={'name': file.name})
+
+
+def validate_file_infection(file):
```
Comment (Jkrzy) [inline crt_portal/cts_forms/validators.py:21]: In testing locally I haven't been able to get this to fail on an upload in excess of 100mb.

The clamav container reports: `^INSTREAM: Size limit reached, (requested: 1024, max: 502)` but the python app isn't responding to the browser request. 

This may be a non-issue once the other, application side, file checks are in place and we're returning an error message on large files before handing to clamav.

File: crt_portal/cts_forms/views.py [hunk 4]
```diff
@@ -630,6 +633,34 @@ def post(self, request):
             return render(request, 'forms/complaint_view/actions/index.html', output)
 
 
+class SaveReportAttachmentView(LoginRequiredMixin, FormView):
```
Comment (Jkrzy) [inline crt_portal/cts_forms/views.py:636]: Opportunity here to specify, with something like [http_method_names](https://docs.djangoproject.com/en/2.2/ref/class-based-views/base/#django.views.generic.base.View.http_method_names), that we're only handling POST requests. So we return a 405 on anything else instead of a 500.

File: crt_portal/cts_forms/validators.py (hunk 3 above)
Comment (timoballard) [reply crt_portal/cts_forms/validators.py:23]: Yes- @ilodidoj 's validation work has been pulled in, so file size is now checked before anything gets sent to ClamAV:

image

File: crt_portal/cts_forms/models.py [hunk 5]
```diff
@@ -357,6 +358,13 @@ def contact_full_name(self):
         return first or last
 
 
+class ReportAttachment(models.Model):
+    file = models.FileField(upload_to='attachments', validators=[validate_file_attachment])
+    filename = models.CharField(max_length=255)
+    user = models.ForeignKey(User, blank=True, null=True, on_delete=models.SET_NULL)
+    report = models.ForeignKey(Report, on_delete=models.CASCADE, related_name='attachments')
+
```
Comment (Jkrzy) [inline crt_portal/cts_forms/models.py:366]: We're appending a datetime stamp to the file names as stored in S3, it may be good to track a `datetime_created` on the model as well.

File: run.sh [hunk 6]
```diff
@@ -1,21 +1,29 @@
...
 node node_modules/gulp/bin/gulp build-sass
-node node_modules/gulp/bin/gulp watch-sass &
 
-echo collect static assets...
-python /code/crt_portal/manage.py collectstatic --noinput
+# If LOCALSTACK is set in environment, this will upload static files to the localstack s3 service running in docker
+# Otherwise the development server is handling static files
+if [[ -n "${USE_LOCALSTACK}" ]]; then
+    aws --endpoint-url=${LOCALSTACK_URL} s3 mb s3://crt-portal
+    aws --endpoint-url=${LOCALSTACK_URL} s3 mb s3://crt-private
```
Comment (Jkrzy) [inline run.sh:15]: This is convenient, thanks!!!

File: README.md [hunk 7]
```diff
@@ -280,6 +280,7 @@ Create postgres DB and S3 with development settings:
 
     cf create-service aws-rds shared-psql crt-db
     cf create-service s3 basic-public crt-s3
+    cf create-service s3 basic sso-creds
```
Comment (timoballard) [inline README.md:283]: I think now that there will be data beyond the SSO creds stored in this bucket, we should rename it to something more generic (e.g., `crt-private`) to avoid confusion. I will document this as a standalone card

File: crt_portal/cts_forms/models.py (hunk 5 above)
Comment (timoballard) [reply crt_portal/cts_forms/models.py:367]: yes!

File: crt_portal/static/js/attachments.js [hunk 8]
```diff
@@ -0,0 +1,17 @@
...
+    event.preventDefault();
+    file_el.click();
+  };
+
+  file_el.onchange = function(event) {
+    var form = document.getElementById('complaint-view-attachments');
+    form.submit();
+
+    var modal = document.getElementById('attachment-uploading--modal');
+    root.CRT.openModal(modal);
```
Comment (Jkrzy) [inline crt_portal/static/js/attachments.js:15]: This opens the modal but generates an exception on modal.js:50 when `openModal` attempts to set the focus to the first focusable element. A byproduct of openModal being implemented for modals with interactions though in this use-case the modal has none.

File: crt_portal/cts_forms/templates/forms/complaint_view/show/attachments.html [hunk 9]
```diff
@@ -0,0 +1,34 @@
...
+    <legend class="usa-sr-only">Complaint detail view attachments</legend>
+    <input type="hidden" value="{{ return_url_args }}" name="next" id="next" />
+    <input type="hidden" value="{{ index }}" name="index" id="index" />
+    <input type="hidden" value="{{ data.pk }}" name="report" id="report_id" />
+    {% for attachment in attachments %}
+      <p>{{ attachment.filename }}</p>
+    {% endfor %}
+    <div class="intake-section">
+      
+      <input type="file" name="file" id="file_input" hidden/>
```
Comment (Jkrzy) [inline crt_portal/cts_forms/templates/forms/complaint_view/show/attachments.html:22]: Opportunity here to specify the filetypes we're accepting on the browser side as well -- https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input/file#limiting_accepted_file_types

File: crt_portal/static/js/attachments.js (hunk 8 above)
Comment (timoballard) [reply crt_portal/static/js/attachments.js:15]: Thanks! Added a hidden input to the attachment modal

File: crt_portal/cts_forms/templates/partials/attachment-uploading-modal.html [hunk 10]
```diff
@@ -0,0 +1,19 @@
...
+      <div class="modal-header">
+        <h1 id="modal_dialog_header" class="h3__display">
+          {% trans "Uploading..." %}
+        </h1>
+      </div>
+      <div class="modal-form">
+        <p>
+          {% trans "Your attachment is uploading..." %}
+        </p>
+        <input type="hidden" />
```
Comment (Jkrzy) [inline crt_portal/cts_forms/templates/partials/attachment-uploading-modal.html:15]: Ahh, what if we used `tabindex` to make the `h1` focusable here instead of adding an input? 

Since openModal will focus on the `h1` when called, a screenreader will reader the Uploading content and give folks non-visual feedback as well as the modal.

Comment (timoballard) [reply crt_portal/cts_forms/templates/partials/attachment-uploading-modal.html:15]: Yes- wasn't thinking about accessibility! Thanks - making this change w/ Imran now

**Review by Jkrzy — APPROVED** [review]
Great first pass for this issue, thanks @ilodidoj and @timoballard

File: README.md (hunk 7 above)
Comment (Jkrzy) [reply README.md:283]: Thanks!
